src/push-state-tree.js

```diff
@@ -56,6 +51,59 @@ const CHANGE = 'change';
 const MATCH = 'match';
 const OLD_MATCH = 'oldMatch';
 
+// Internal history keep tracking of all changes in the location history, it can be reset any
+let internalHistory;
+function InternalLocation(id, url) {
+  this.id = id;
+  this.url = url;
+}
+function InternalHistory() {
+  if (!(this instanceof InternalHistory)) {
+    return internalHistory = new InternalHistory();
+  }
+
+  // It's a object like an array, but it can start from a certain point to allow memory cleanup
+  if (internalHistory && internalHistory.length) {
+    this.startFrom = internalHistory.length - 1;
+    let savedLocation = internalHistory[this.startFrom];
+
+    // Break the link to previous to allow GC collect unused values
+    if (savedLocation) savedLocation.previous = undefined;
+
+    this[this.startFrom] = savedLocation;
+    this.length = internalHistory.length;
```

Comment:
Actually it's kind of a factory, accessing the old instance to copy the reference value, them it generates the new instance on `this` and let the old instance without reference in the memory to the GC collect at some point... However we will probably get rid of this in the next PR, I think you're right, it's confusing to read.